source/dub/packagemanager.d

```diff
@@ -1151,32 +1160,40 @@ symlink_exit:
 	 * However, due to it being a filesystem interaction, it is managed
 	 * from the `PackageManager`.
 	 *
-	 * Under normal conditions, either `null` is returned (if no selection
-	 * file exists or parsing encountered an error), or a `SelectionFile`
-	 * is returned. Note that the returned `SelectionsFile` might use an
-	 * unsupported version (see `SelectionsFile` documentation).
-	 *
-	 * Note that this is currently not part of the public API, and won't be
-	 * until there is a clear need for it.
-	 *
 	 * Params:
-	 *   project = The `Package` for which to load a selection file.
+	 *   absProjectPath = The absolute path to the root package/project for
+	 *                    which to load the selections file.
 	 *
 	 * Returns:
-	 *   The parsed `SelectionsFile`, or a null instance if none exists.
+	 *   Either `null` (if no selections file exists or parsing encountered an error),
+	 *   or a `SelectionsFileLookupResult`. Note that the nested `SelectionsFile`
+	 *   might use an unsupported version (see `SelectionsFile` documentation).
 	 */
-	package final Nullable!SelectionsFile readSelections(in Package project) {
+	Nullable!SelectionsFileLookupResult readSelections(in NativePath absProjectPath)
+	in (absProjectPath.absolute) {
 		import dub.internal.configy.Read;
 
-		const path = (project.path ~ "dub.selections.json");
-		if (!this.existsFile(path))
-			return typeof(return).init;
-		const content = this.readText(path);
-		// TODO: Remove `StrictMode.Warn` after v1.40 release
-		// The default is to error, but as the previous parser wasn't
-		// complaining, we should first warn the user.
-		return wrapException(parseConfigString!SelectionsFile(
-			content, path.toNativeString(), StrictMode.Warn));
+		alias N = typeof(return);
+
+		// check for dub.selections.json in root project dir first, then walk up its
+		// parent directories and look for inheritable dub.selections.json files
+		for (NativePath dir = absProjectPath; !dir.empty; dir = dir.hasParentPath ? dir.parentPath : NativePath.init) {
+			const path = dir ~ "dub.selections.json";
+			if (!this.existsFile(path))
+				continue;
```

Comment:
Logically this would be better represented as something that returns a path that is then acted on. The code as-is does not mean to be changed (it works, that's what matters) but a loop that ends up with an unconditional `break` is not a loop (or IOW, contains too much code).